src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java

```diff
@@ -190,10 +208,142 @@ public ResultSet getClientInfoProperties() throws SQLException {
         throw new SQLFeatureNotSupportedException("TODO");
     }
 
+    private MongoJsonSchema getFunctionJsonSchema() {
+        MongoJsonSchema schema = new MongoJsonSchema();
+        schema.bsonType = "object";
+        schema.additionalProperties = false;
+        schema.required = new HashSet<String>();
+        schema.properties = new HashMap<String, MongoJsonSchema>();
+
+        schema.required.add(FUNCTION_CAT);
+        schema.required.add(FUNCTION_SCHEM);
+        schema.required.add(FUNCTION_NAME);
+        schema.required.add(REMARKS);
+        schema.required.add(FUNCTION_TYPE);
+        schema.required.add(SPECIFIC_NAME);
+
+        schema.properties.put(FUNCTION_CAT, stringSchema);
+        schema.properties.put(FUNCTION_SCHEM, stringSchema);
+        schema.properties.put(FUNCTION_NAME, stringSchema);
+        schema.properties.put(REMARKS, stringSchema);
+        schema.properties.put(FUNCTION_TYPE, intSchema);
+        schema.properties.put(SPECIFIC_NAME, stringSchema);
+        return schema;
+    }
+
+    private BsonDocument getFunctionValuesDoc(String functionName, String remarks) {
+        BsonDocument root = new BsonDocument();
+        BsonDocument bot = new BsonDocument();
+        root.put("", bot);
+        bot.put(FUNCTION_CAT, new BsonString("def"));
+        bot.put(FUNCTION_SCHEM, new BsonNull());
+        bot.put(FUNCTION_NAME, new BsonString(functionName));
+        bot.put(REMARKS, new BsonString(remarks));
+        bot.put(FUNCTION_TYPE, new BsonInt32(functionNoTable));
+        bot.put(SPECIFIC_NAME, new BsonString(functionName));
+        return root;
+    }
+
     @Override
     public ResultSet getFunctions(String catalog, String schemaPattern, String functionNamePattern)
             throws SQLException {
-        throw new SQLFeatureNotSupportedException("TODO");
+        ArrayList<BsonDocument> docs = new ArrayList<>(MongoFunction.mongoSQLFunctionNames.length);
+        MongoJsonSchema schema = getFunctionJsonSchema();
+
+        Pattern functionPatternRE = null;
+        if (functionNamePattern != null) {
+            functionPatternRE = Pattern.compile(functionNamePattern.replaceAll("%", ".*"));
+        }
+
+        for (MongoFunction func : MongoFunction.mongoSQLFunctions) {
+            String functionName = func.name;
+            String remarks = func.comment;
+            if (functionPatternRE != null && !functionPatternRE.matcher(functionName).matches()) {
+                continue;
+            }
+            BsonDocument doc = getFunctionValuesDoc(func.name, func.comment);
+            docs.add(doc);
+        }
+
+        return new MongoSQLResultSet(null, null);
+        // return MongoSQLResultSet(null, new MongoSQLExplicitCursor(docs), schema);
```

Comment:
What is your preference for the name: my current PR for 551 has `BsonExplicitCursor` since it operates over `BsonDocument`s. Here, you have `MongoSQLExplicitCursor` which also makes sense since it will be used by the MongoSQL implementations of the driver interfaces. I'm leaning a bit more `Bson` than `MongoSQL` since this cursor type doesn't strictly need to be used with just the MongoSQL code, but will go with whichever you prefer.

Reply:
Oh, I'm fine with `Bson`, I just forgot to recheck your PR!